File: app-windows/RunnerForge/Services/ProcessRunner.cs
using System.Diagnostics;
using System.Text;

namespace RunnerForge.Services;

public sealed record ProcessResult(int ExitCode, string StandardOutput, string StandardError)
{
    public bool Succeeded => ExitCode == 0;

    /// <summary>Both streams, for error messages that should show everything.</summary>
    public string CombinedOutput =>
        string.Join(Environment.NewLine,
            new[] { StandardOutput, StandardError }.Where(s => !string.IsNullOrWhiteSpace(s)));
}

/// <summary>
/// Runs external processes and streams their output into the LogBus.
/// </summary>
/// <remarks>
/// Secrets are passed through <paramref name="environment"/> or standard input,
/// NEVER through arguments: an argument list is readable by any user on the
/// machine via the process list. There is deliberately no overload that takes a
/// secret as an argument.
/// </remarks>
public sealed class ProcessRunner(LogBus logBus)
{
    private readonly LogBus _logBus = logBus;

    public async Task<ProcessResult> RunAsync(
        string fileName,
        IEnumerable<string> arguments,
        string? workingDirectory = null,
        IReadOnlyDictionary<string, string>? environment = null,
        string? standardInput = null,
        string logSource = "process",
        bool streamOutput = true,
        CancellationToken cancellationToken = default)
    {
        var startInfo = new ProcessStartInfo
        {
            FileName = fileName,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            RedirectStandardInput = standardInput is not null,
            UseShellExecute = false,
            CreateNoWindow = true,
            WorkingDirectory = workingDirectory ?? Environment.CurrentDirectory,

            // Decode child output as UTF-8 rather than the console's OEM code
            // page. Without this, a localised Windows mangles every non-ASCII
            // character on its way into the log: a real run produced
            //     "found: Firma de c\u00a2digo"
            // for "Firma de c\u00f3digo", because CP850 was being read as if it
            // were the default encoding. A log that corrupts the message is
            // worse than one that omits it, since the reader cannot tell which
            // happened.
            StandardOutputEncoding = System.Text.Encoding.UTF8,
            StandardErrorEncoding = System.Text.Encoding.UTF8,
        };

        foreach (string argument in arguments)
        {
            startInfo.ArgumentList.Add(argument);
        }

        if (environment is not null)
        {
            foreach ((string key, string value) in environment)
            {
                startInfo.Environment[key] = value;
                // Anything injected as an environment variable is, by
                // definition, something we do not want echoed back at us.
                _logBus.RegisterSecret(value);
            }
        }

        // The argument list is logged, which is safe precisely because secrets
        // are never in it. Redaction still runs as a second line of defence.
        _logBus.Debug(logSource, $"$ {fileName} {string.Join(' ', startInfo.ArgumentList)}");

        using var process = new Process { StartInfo = startInfo, EnableRaisingEvents = true };

        var stdout = new StringBuilder();
        var stderr = new StringBuilder();

        process.OutputDataReceived += (_, e) =>
        {
            if (e.Data is null) return;
            stdout.AppendLine(e.Data);
            if (streamOutput) _logBus.Info(logSource, e.Data);
        };

        process.ErrorDataReceived += (_, e) =>
        {
            if (e.Data is null) return;
            stderr.AppendLine(e.Data);
            if (streamOutput) _logBus.Warning(logSource, e.Data);
        };

        try
        {
            process.Start();
        }
        catch (Exception ex)
        {
            _logBus.Error(logSource, $"could not start '{fileName}': {ex.Message}");
            return new ProcessResult(-1, "", ex.Message);
        }

        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        if (standardInput is not null)
        {
            // The JIT config arrives this way. It is written, flushed, and the
            // stream closed immediately; the value never touches a file.
            _logBus.RegisterSecret(standardInput);
            await process.StandardInput.WriteLineAsync(standardInput.AsMemory(), cancellationToken)
                                       .ConfigureAwait(false);
            await process.StandardInput.FlushAsync(cancellationToken).ConfigureAwait(false);
            process.StandardInput.Close();
        }

        try
        {
            await process.WaitForExitAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            TryKill(process, logSource);
            throw;
        }

        return new ProcessResult(process.ExitCode, stdout.ToString(), stderr.ToString());
    }

    private void TryKill(Process process, string logSource)
    {
        try
        {
            if (!process.HasExited)
            {
                _logBus.Warning(logSource, $"cancelled; killing pid {process.Id}");
                process.Kill(entireProcessTree: true);
            }
        }
        catch (Exception ex)
        {
            _logBus.Warning(logSource, $"could not kill the process: {ex.Message}");
        }
    }

    /// <summary>True when the executable can be found on PATH.</summary>
    public static bool IsOnPath(string executable)
    {
        string? pathValue = Environment.GetEnvironmentVariable("PATH");
        if (string.IsNullOrEmpty(pathValue)) return false;

        string[] extensions = OperatingSystem.IsWindows()
            ? (Environment.GetEnvironmentVariable("PATHEXT") ?? ".EXE;.CMD;.BAT").Split(';')
            : [""];

        foreach (string directory in pathValue.Split(Path.PathSeparator))
        {
            if (string.IsNullOrWhiteSpace(directory)) continue;
            foreach (string extension in extensions)
            {
                try
                {
                    if (File.Exists(Path.Combine(directory, executable + extension))) return true;
                }
                catch (ArgumentException)
                {
                    // A malformed PATH entry is not worth failing over.
                }
            }
        }

        return false;
    }
}

File: app-macos/Sources/RunnerForge/Services/ProcessRunner.swift
import Foundation

public struct ProcessResult: Sendable {
    public let exitCode: Int32
    public let standardOutput: String
    public let standardError: String

    public var succeeded: Bool { exitCode == 0 }

    /// Both streams, for error messages that should show everything.
    public var combinedOutput: String {
        [standardOutput, standardError]
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .joined(separator: "\n")
    }
}

/// Runs external processes and streams their output into the LogBus.
///
/// Secrets are passed through `environment` or standard input, NEVER through
/// arguments: an argument list is readable by any user on the machine via the
/// process list. There is deliberately no overload that takes a secret as an
/// argument.
public struct ProcessRunner: Sendable {
    private let logBus: LogBus

    public init(logBus: LogBus) { self.logBus = logBus }

    public func run(
        _ executable: String,
        arguments: [String],
        workingDirectory: String? = nil,
        environment: [String: String]? = nil,
        standardInput: String? = nil,
        logSource: String = "process",
        streamOutput: Bool = true
    ) async throws -> ProcessResult {
        let process = Process()
        process.executableURL = Self.resolve(executable)
        process.arguments = arguments

        if let workingDirectory {
            process.currentDirectoryURL = URL(fileURLWithPath: workingDirectory)
        }

        if let environment {
            var merged = ProcessInfo.processInfo.environment
            for (key, value) in environment {
                merged[key] = value
                // Anything injected as an environment variable is, by definition,
                // something we do not want echoed back at us.
                await logBus.registerSecret(value)
            }
            process.environment = merged
        }

        let outputPipe = Pipe()
        let errorPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = errorPipe

        let inputPipe: Pipe? = standardInput != nil ? Pipe() : nil
        if let inputPipe { process.standardInput = inputPipe }

        // The argument list is logged, which is safe precisely because secrets
        // are never in it. Redaction still runs as a second line of defence.
        await logBus.debug(logSource, "$ \(executable) \(arguments.joined(separator: " "))")

        try process.run()

        if let inputPipe, let standardInput {
            // The JIT config arrives this way. It is written, flushed, and the
            // handle closed immediately; the value never touches a file.
            await logBus.registerSecret(standardInput)
            inputPipe.fileHandleForWriting.write(Data((standardInput + "\n").utf8))
            try? inputPipe.fileHandleForWriting.close()
        }

        let outputData = try outputPipe.fileHandleForReading.readToEnd() ?? Data()
        let errorData = try errorPipe.fileHandleForReading.readToEnd() ?? Data()

        process.waitUntilExit()

        let output = String(decoding: outputData, as: UTF8.self)
        let errorOutput = String(decoding: errorData, as: UTF8.self)

        if streamOutput {
            for line in output.split(separator: "\n", omittingEmptySubsequences: true) {
                await logBus.info(logSource, String(line))
            }
            for line in errorOutput.split(separator: "\n", omittingEmptySubsequences: true) {
                await logBus.warning(logSource, String(line))
            }
        }

        return ProcessResult(
            exitCode: process.terminationStatus,
            standardOutput: output,
            standardError: errorOutput)
    }

    /// Resolves a bare name against PATH, or uses the path as given.
    static func resolve(_ executable: String) -> URL {
        if executable.contains("/") { return URL(fileURLWithPath: executable) }

        let searchPath = ProcessInfo.processInfo.environment["PATH"] ?? "/usr/bin:/bin:/usr/local/bin"
        for directory in searchPath.split(separator: ":") {
            let candidate = URL(fileURLWithPath: String(directory)).appendingPathComponent(executable)
            if FileManager.default.isExecutableFile(atPath: candidate.path) { return candidate }
        }

        return URL(fileURLWithPath: "/usr/bin/env")
    }

    /// True when the executable can be found on PATH.
    public static func isOnPath(_ executable: String) -> Bool {
        let searchPath = ProcessInfo.processInfo.environment["PATH"] ?? "/usr/bin:/bin:/usr/local/bin"
        for directory in searchPath.split(separator: ":") {
            let candidate = URL(fileURLWithPath: String(directory)).appendingPathComponent(executable)
            if FileManager.default.isExecutableFile(atPath: candidate.path) { return true }
        }
        return false
    }
}
